spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java

```diff
@@ -241,24 +272,32 @@ public void setupListenerContainer(MessageListenerContainer listenerContainer, M
 	protected abstract MessageListener<K, V> createMessageListener(MessageListenerContainer container,
 			MessageConverter messageConverter);
 
+	@SuppressWarnings("unchecked")
 	private void setupMessageListener(MessageListenerContainer container, MessageConverter messageConverter) {
-		MessageListener<K, V> messageListener = createMessageListener(container, messageConverter);
+		Object messageListener = createMessageListener(container, messageConverter);
 		Assert.state(messageListener != null, "Endpoint [" + this + "] must provide a non null message listener");
-		if (this.recordFilterStrategy != null) {
+		if (this.retryTemplate != null) {
 			if (messageListener instanceof AcknowledgingMessageListener) {
-				@SuppressWarnings("unchecked")
-				AcknowledgingMessageListener<K, V> aml = (AcknowledgingMessageListener<K, V>) messageListener;
-				aml = new FilteringAcknowledgingMessageListenerAdapter<>(this.recordFilterStrategy, aml, this.ackDiscarded);
-				container.setupMessageListener(aml);
+				messageListener = new RetryingAcknowledgingMessageListenerAdapter<>(
+						(AcknowledgingMessageListener<K, V>) messageListener, this.retryTemplate,
+						this.recoveryCallback);
 			}
 			else {
-				messageListener = new FilteringMessageListenerAdapter<>(this.recordFilterStrategy, messageListener);
-				container.setupMessageListener(messageListener);
+				messageListener = new FilteringMessageListenerAdapter<>(this.recordFilterStrategy,
+						(MessageListener<K, V>) messageListener);
 			}
 		}
-		else {
-			container.setupMessageListener(messageListener);
+		if (this.recordFilterStrategy != null) {
+			if (messageListener instanceof AcknowledgingMessageListener) {
+				messageListener = new FilteringAcknowledgingMessageListenerAdapter<>(this.recordFilterStrategy,
+						(AcknowledgingMessageListener<K, V>) messageListener, this.ackDiscarded);
```

Comment:
Can't we revise API and really have `messageListener` as a first ctor arg?
Just for cleaner code :smile: